File: backend/utils/unipile.py
"""
Unipile API client for the LinkedPilot backend worker.

Wraps all Unipile REST endpoints needed by the action queue dispatcher.
Reads credentials from environment variables (UNIPILE_DSN, UNIPILE_API_KEY).
"""
import os
from typing import Optional

import httpx
from dotenv import load_dotenv

from utils.logger import logger

load_dotenv()


class UnipileClient:
    """Async HTTP client for the Unipile REST API."""

    def __init__(self):
        dsn = os.getenv("UNIPILE_DSN")
        api_key = os.getenv("UNIPILE_API_KEY")

        if not dsn:
            raise RuntimeError(
                "[Unipile] UNIPILE_DSN is not set. "
                "Add it to your .env file (see .env.example)."
            )
        if not api_key:
            raise RuntimeError(
                "[Unipile] UNIPILE_API_KEY is not set. "
                "Add it to your .env file (see .env.example)."
            )

        self.base_url = f"https://{dsn}/api/v1"
        self._headers = {
            "X-API-KEY": api_key,
            "Content-Type": "application/json",
            "Accept": "application/json",
        }

    # -------------------------------------------------------------------------
    # Core HTTP dispatcher
    # -------------------------------------------------------------------------

    async def _request(self, method: str, path: str, **kwargs) -> dict:
        """
        Central HTTP dispatcher.

        Raises httpx.HTTPStatusError on non-2xx responses so callers can map
        status codes to the appropriate exception type.
        """
        url = f"{self.base_url}{path}"
        # Remove Content-Type for multipart requests (httpx sets it automatically)
        headers = dict(self._headers)
        if "files" in kwargs or "data" in kwargs:
            headers.pop("Content-Type", None)

        async with httpx.AsyncClient(timeout=60.0) as client:
            resp = await client.request(method, url, headers=headers, **kwargs)

        try:
            resp.raise_for_status()
        except httpx.HTTPStatusError as exc:
            logger.error(
                f"[Unipile] {method} {path} → HTTP {exc.response.status_code}: "
                f"{exc.response.text[:500]}"
            )
            raise

        return resp.json()

    # -------------------------------------------------------------------------
    # Account management
    # -------------------------------------------------------------------------

    async def list_accounts(self) -> list:
        """GET /api/v1/accounts — list all connected accounts."""
        data = await self._request("GET", "/accounts")
        return data.get("items", [])

    async def resync_account(self, unipile_account_id: str) -> dict:
        """POST /api/v1/accounts/{id}/resync — reconnect a disconnected account."""
        return await self._request("POST", f"/accounts/{unipile_account_id}/resync")

    async def delete_account(self, unipile_account_id: str) -> dict:
        """DELETE /api/v1/accounts/{id} — remove an account from Unipile."""
        return await self._request("DELETE", f"/accounts/{unipile_account_id}")

    # -------------------------------------------------------------------------
    # Profiles
    # -------------------------------------------------------------------------

    async def get_profile(self, account_id: str, identifier: str) -> dict:
        """
        GET /api/v1/users/{identifier}?account_id=...
        Retrieve a LinkedIn user profile by public identifier or provider ID.
        """
        return await self._request(
            "GET",
            f"/users/{identifier}",
            params={"account_id": account_id},
        )

    async def get_own_profile(self, account_id: str) -> dict:
        """GET /api/v1/users/me?account_id=... — retrieve the authenticated user's profile."""
        return await self._request("GET", "/users/me", params={"account_id": account_id})

    async def get_company_profile(self, account_id: str, identifier: str) -> dict:
        """GET /api/v1/users/company/{identifier}?account_id=..."""
        return await self._request(
            "GET",
            f"/users/company/{identifier}",
            params={"account_id": account_id},
        )

    async def get_relations(self, account_id: str, cursor: Optional[str] = None) -> dict:
        """GET /api/v1/users/relations?account_id=... — list connections."""
        params = {"account_id": account_id}
        if cursor:
            params["cursor"] = cursor
        return await self._request("GET", "/users/relations", params=params)

    # -------------------------------------------------------------------------
    # Messaging
    # -------------------------------------------------------------------------

    async def list_chats(self, account_id: str, limit: int = 50) -> dict:
        """GET /api/v1/chats?account_id=..."""
        return await self._request(
            "GET",
            "/chats",
            params={"account_id": account_id, "limit": limit},
        )

    async def get_chat(self, chat_id: str) -> dict:
        """GET /api/v1/chats/{id}"""
        return await self._request("GET", f"/chats/{chat_id}")

    async def list_messages(self, chat_id: str) -> dict:
        """GET /api/v1/chats/{id}/messages"""
        return await self._request("GET", f"/chats/{chat_id}/messages")

    async def list_attendees(self, chat_id: str) -> dict:
        """GET /api/v1/chats/{id}/attendees"""
        return await self._request("GET", f"/chats/{chat_id}/attendees")

    async def get_or_create_chat(self, account_id: str, attendee_id: str) -> str:
        """
        Silently resolve a chat ID for the given attendee.

        1. Search existing chats for one that contains the attendee.
        2. If not found, create a new chat (POST /api/v1/chats).
        Returns the chat_id string.
        """
        # Search existing chats
        try:
            chats_data = await self.list_chats(account_id, limit=100)
            for chat in chats_data.get("items", []):
                attendees = chat.get("attendees", [])
                for att in attendees:
                    if att.get("provider_id") == attendee_id or att.get("id") == attendee_id:
                        return chat["id"]
        except httpx.HTTPStatusError:
            pass  # Fall through to create

        # Create new chat
        result = await self._request(
            "POST",
            "/chats",
            json={"account_id": account_id, "attendees_ids": [attendee_id]},
        )
        return result["id"]

    async def send_message(self, chat_id: str, text: str, account_id: str) -> dict:
        """POST /api/v1/chats/{id}/messages"""
        return await self._request(
            "POST",
            f"/chats/{chat_id}/messages",
            json={"text": text, "account_id": account_id},
        )

    async def start_new_chat(self, account_id: str, attendee_id: str, text: str) -> dict:
        """POST /api/v1/chats — start a new DM conversation."""
        return await self._request(
            "POST",
            "/chats",
            json={
                "account_id": account_id,
                "attendees_ids": [attendee_id],
                "text": text,
            },
        )

    async def send_inmail(
        self,
        account_id: str,
        attendee_id: str,
        text: str,
        api: str = "classic",
    ) -> dict:
        """POST /api/v1/chats with inmail:true — send an InMail."""
        return await self._request(
            "POST",
            "/chats",
            json={
                "account_id": account_id,
                "attendees_ids": [attendee_id],
                "text": text,
                "options": {"linkedin": {"api": api, "inmail": True}},
            },
        )

    # -------------------------------------------------------------------------
    # Invitations
    # -------------------------------------------------------------------------

    async def send_invitation(
        self,
        account_id: str,
        provider_id: str,
        message: Optional[str] = None,
    ) -> dict:
        """POST /api/v1/users/invite — send a LinkedIn connection request."""
        body: dict = {"account_id": account_id, "provider_id": provider_id}
        if message:
            body["message"] = message
        return await self._request("POST", "/users/invite", json=body)

    async def list_sent_invitations(self, account_id: str) -> dict:
        """GET /api/v1/users/invitations/sent?account_id=..."""
        return await self._request(
            "GET",
            "/users/invitations/sent",
            params={"account_id": account_id},
        )

    async def cancel_invitation(self, account_id: str, invitation_id: str) -> dict:
        """DELETE /api/v1/users/invitations/{id}"""
        return await self._request(
            "DELETE",
            f"/users/invitations/{invitation_id}",
            params={"account_id": account_id},
        )

    async def list_received_invitations(self, account_id: str) -> dict:
        """GET /api/v1/users/invitations/received?account_id=..."""
        return await self._request(
            "GET",
            "/users/invitations/received",
            params={"account_id": account_id},
        )

    async def handle_invitation(
        self,
        account_id: str,
        invitation_id: str,
        action: str,  # "accept" or "decline"
    ) -> dict:
        """POST /api/v1/users/invitations/handle — accept or decline a received invitation."""
        return await self._request(
            "POST",
            "/users/invitations/handle",
            json={
                "account_id": account_id,
                "invitation_id": invitation_id,
                "action": action,
            },
        )

    # -------------------------------------------------------------------------
    # Search
    # -------------------------------------------------------------------------

    async def search(self, account_id: str, params: dict) -> dict:
        """
        POST /api/v1/linkedin/search?account_id=...
        Supports Classic, Sales Navigator, and Recruiter tiers.
        params should include: api_type, category, keywords, filters, cursor, etc.
        """
        return await self._request(
            "POST",
            "/linkedin/search",
            params={"account_id": account_id},
            json=params,
        )

    async def get_search_parameters(
        self,
        account_id: str,
        param_type: str,
        keywords: Optional[str] = None,
    ) -> dict:
        """GET /api/v1/linkedin/search/parameters — fetch filter options."""
        query_params = {"account_id": account_id, "type": param_type}
        if keywords:
            query_params["keywords"] = keywords
        return await self._request(
            "GET",
            "/linkedin/search/parameters",
            params=query_params,
        )

    # -------------------------------------------------------------------------
    # Posts & Engagement
    # -------------------------------------------------------------------------

    async def list_user_posts(self, account_id: str, identifier: str) -> dict:
        """GET /api/v1/users/{identifier}/posts?account_id=..."""
        return await self._request(
            "GET",
            f"/users/{identifier}/posts",
            params={"account_id": account_id},
        )

    async def get_post(self, account_id: str, post_id: str) -> dict:
        """GET /api/v1/posts/{id}?account_id=..."""
        return await self._request(
            "GET",
            f"/posts/{post_id}",
            params={"account_id": account_id},
        )

    async def create_post(self, account_id: str, text: str) -> dict:
        """POST /api/v1/posts — create a new LinkedIn post."""
        return await self._request(
            "POST",
            "/posts",
            json={"account_id": account_id, "text": text},
        )

    async def comment_on_post(self, account_id: str, post_id: str, text: str) -> dict:
        """POST /api/v1/posts/{id}/comments"""
        return await self._request(
            "POST",
            f"/posts/{post_id}/comments",
            json={"account_id": account_id, "text": text},
        )

    async def list_post_comments(self, account_id: str, post_id: str) -> dict:
        """GET /api/v1/posts/{id}/comments?account_id=..."""
        return await self._request(
            "GET",
            f"/posts/{post_id}/comments",
            params={"account_id": account_id},
        )

    async def react_to_post(
        self,
        account_id: str,
        post_id: str,
        reaction_type: str = "LIKE",
    ) -> dict:
        """POST /api/v1/posts/{id}/reactions"""
        return await self._request(
            "POST",
            f"/posts/{post_id}/reactions",
            json={"account_id": account_id, "reaction_type": reaction_type},
        )

File: src/lib/unipile.js
/**
 * Unipile API client — all calls proxied through the backend worker.
 *
 * The backend exposes:
 *   POST /api/proxy          — generic proxy for any Unipile endpoint
 *   POST /api/linkedin/search
 *   GET  /api/linkedin/search/parameters
 *
 * This keeps VITE_UNIPILE_DSN and VITE_UNIPILE_API_KEY off the browser
 * network tab entirely. The frontend only ever talks to the local backend.
 */

const BACKEND = () => import.meta.env.VITE_BACKEND_URL || 'http://localhost:3000'

/** Generic proxy call */
async function proxy({ method = 'GET', path, account_id, params, body } = {}) {
  const url = `${BACKEND()}/api/proxy`
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
    body: JSON.stringify({ method, path, account_id, params, body }),
  })
  if (!res.ok) {
    let b = {}
    try { b = await res.json() } catch { /* ignore */ }
    const err = new Error(b?.error ?? b?.message ?? `Proxy error ${res.status}`)
    err.response = { status: res.status, data: b }
    throw err
  }
  if (res.status === 204) return {}
  return res.json()
}

/** Build query string, skipping null/undefined values */
function qs(params) {
  const p = new URLSearchParams()
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== null) p.set(k, String(v))
  }
  const s = p.toString()
  return s ? `?${s}` : ''
}

export const unipile = {

  // ── account ──────────────────────────────────────────────────────────────
  account: {
    getAll: ({ limit, cursor } = {}) =>
      proxy({ path: `/accounts${qs({ limit, cursor })}` }),

    getOne: (account_id) =>
      proxy({ path: `/accounts/${account_id}` }),

    connectLinkedin: ({ username, password, country, proxy: p } = {}) =>
      proxy({ method: 'POST', path: '/accounts', body: { provider: 'LINKEDIN', username, password, ...(country ? { country } : {}), ...(p ? { proxy: p } : {}) } }),

    connectLinkedinWithCookie: ({ access_token, user_agent } = {}) =>
      proxy({ method: 'POST', path: '/accounts', body: { provider: 'LINKEDIN', access_token, ...(user_agent ? { user_agent } : {}) } }),

    reconnect: ({ account_id, ...rest } = {}) =>
      proxy({ method: 'POST', path: `/accounts/${account_id}`, body: { account_id, ...rest } }),

    delete: (account_id) =>
      proxy({ method: 'DELETE', path: `/accounts/${account_id}` }),

    solveCodeCheckpoint: ({ account_id, provider = 'LINKEDIN', code } = {}) =>
      proxy({ method: 'POST', path: '/accounts/checkpoint', body: { account_id, provider, code } }),

    resendCheckpointNotification: ({ account_id, provider = 'LINKEDIN' } = {}) =>
      proxy({ method: 'POST', path: '/accounts/checkpoint/resend', body: { account_id, provider } }),

    createHostedAuthLink: (params = {}) =>
      proxy({ method: 'POST', path: '/hosted/accounts/link', body: params }),

    resyncLinkedinAccount: ({ account_id } = {}) =>
      proxy({ path: `/accounts/${account_id}/sync` }),
  },

  // ── messaging ─────────────────────────────────────────────────────────────
  messaging: {
    getAllChats: ({ account_id, account_type, limit, cursor, unread } = {}) =>
      proxy({ path: `/chats${qs({ account_id, account_type, limit, cursor, unread })}` }),

    getChat: (chat_id) =>
      proxy({ path: `/chats/${chat_id}` }),

    getAllMessagesFromChat: ({ chat_id, limit, cursor } = {}) =>
      proxy({ path: `/chats/${chat_id}/messages${qs({ limit, cursor })}` }),

    getAllAttendeesFromChat: (chat_id) =>
      proxy({ path: `/chats/${chat_id}/attendees` }),

    getAllAttendees: ({ account_id, limit, cursor } = {}) =>
      proxy({ path: `/chat_attendees${qs({ account_id, limit, cursor })}` }),

    // sendMessage and startNewChat use FormData — proxied as JSON body, backend handles
    sendMessage: ({ chat_id, text } = {}) =>
      proxy({ method: 'POST', path: `/chats/${chat_id}/messages`, body: { text } }),

    startNewChat: ({ account_id, attendees_ids = [], text, subject, options } = {}) =>
      proxy({ method: 'POST', path: '/chats', account_id, body: { account_id, attendees_ids, text, subject, options } }),

    setChatStatus: ({ chat_id, action, value } = {}) =>
      proxy({ method: 'PATCH', path: `/chats/${chat_id}`, body: { action, value } }),
  },

  // ── users ─────────────────────────────────────────────────────────────────
  users: {
    getProfile: ({ account_id, identifier, linkedin_sections, linkedin_api } = {}) => {
      const sections = Array.isArray(linkedin_sections) ? linkedin_sections.join(',') : linkedin_sections
      return proxy({ path: `/users/${encodeURIComponent(identifier)}${qs({ account_id, linkedin_sections: sections, linkedin_api })}` })
    },

    getOwnProfile: (account_id) =>
      proxy({ path: `/users/me${qs({ account_id })}` }),

    getCompanyProfile: ({ account_id, identifier } = {}) =>
      proxy({ path: `/linkedin/company/${encodeURIComponent(identifier)}${qs({ account_id })}` }),

    getAllRelations: ({ account_id, limit, cursor } = {}) =>
      proxy({ path: `/users/relations${qs({ account_id, limit, cursor })}` }),

    sendInvitation: ({ account_id, provider_id, message } = {}) =>
      proxy({ method: 'POST', path: '/users/invite', body: { account_id, provider_id, ...(message ? { message } : {}) } }),

    getAllInvitationsSent: ({ account_id, limit, cursor } = {}) =>
      proxy({ path: `/users/invite/sent${qs({ account_id, limit, cursor })}` }),

    cancelInvitationSent: ({ account_id, invitation_id } = {}) =>
      proxy({ method: 'DELETE', path: `/users/invite/sent/${invitation_id}${qs({ account_id })}` }),

    getAllInvitationsReceived: ({ account_id, limit, cursor } = {}) =>
      proxy({ path: `/users/invite/received${qs({ account_id, limit, cursor })}` }),

    handleInvitation: ({ account_id, invitation_id, action, linkedin_token } = {}) =>
      proxy({
        method: 'POST',
        path: `/users/invite/received/${invitation_id}`,
        body: { account_id, action, ...(linkedin_token ? { linkedin_token } : {}) },
      }),

    getAllPosts: ({ account_id, identifier, limit, cursor, is_company } = {}) =>
      proxy({ path: `/users/${encodeURIComponent(identifier)}/posts${qs({ account_id, limit, cursor, is_company })}` }),

    getPost: ({ account_id, post_id } = {}) =>
      proxy({ path: `/posts/${encodeURIComponent(post_id)}${qs({ account_id })}` }),

    createPost: ({ account_id, text } = {}) =>
      proxy({ method: 'POST', path: '/posts', body: { account_id, text } }),

    getAllPostComments: ({ account_id, post_id, limit, cursor } = {}) =>
      proxy({ path: `/posts/${encodeURIComponent(post_id)}/comments${qs({ account_id, limit, cursor })}` }),

    sendPostComment: ({ account_id, post_id, text } = {}) =>
      proxy({ method: 'POST', path: `/posts/${encodeURIComponent(post_id)}/comments`, body: { account_id, text } }),

    sendPostReaction: ({ account_id, post_id, reaction_type = 'LIKE' } = {}) =>
      proxy({ method: 'POST', path: '/posts/reaction', body: { account_id, post_id, reaction_type } }),
  },
}
